### PR TITLE
Kahlia Age 9 artifacts - syncom, phillips, marilyn, and luna

### PR DESCRIPTION
**Pull request by ultimatefiend**
-random fix for hunter-lenox
-4 age 9 cards

### CODE REVIEW

File: innovation.game.php [hunk 1]
```diff
@@ -12664,6 +12787,62 @@ function stInterInteractionStep() {
...
+                        $color_detect[$card['color']] = 1; // flag the color
+                        $card_array[] = $card;
+                    }
+                    
+                    // If you revealed all five colors, you win.
+                    if ($repeat_colors == false) {
+                        self::notifyPlayer($player_id, 'log', clienttranslate('${You} revealed all 5 colors.'), array('You' => 'You'));
+                        self::notifyAllPlayersBut($player_id, 'log', clienttranslate('${player_name} revealed all 5 colors.'), array('player_name' => self::getColoredText(self::getPlayerNameFromId($player_id), $player_id)));
+                        self::setGameStateValue('winner_by_dogma', $player_id);
+                        self::trace('EOG bubbled from self::stPlayerInvolvedTurn Syncom 3');
```
Comment (micahstairs) [inline innovation.game.php:12826]: This code is in stInterInteractionStep()

self::trace('EOG bubbled from self::stInterInteractionStep Syncom 3');

File: innovation.game.php [hunk 2]
```diff
@@ -12664,6 +12787,62 @@ function stInterInteractionStep() {
...
+                   
+                    break;                    
+                    
+                case "204N1B":
+                    // "If you have exactly 25 points, you win."
+                    if (self::getPlayerScore($player_id) == 25) {
+                        self::notifyPlayer($player_id, 'log', clienttranslate('${You} have 25 points exactly.'), array('You' => 'You'));
+                        self::notifyAllPlayersBut($player_id, 'log', clienttranslate('${player_name} has 25 points exactly.'), array('player_name' => self::getColoredText(self::getPlayerNameFromId($player_id), $player_id)));
+                        self::setGameStateValue('winner_by_dogma', $player_id);
+                        self::trace('EOG bubbled from self::stPlayerInvolvedTurn Marilyn Diptych');
```
Comment (micahstairs) [inline innovation.game.php:12842]: stInterInteractionStep

File: innovation.game.php [hunk 3]
```diff
@@ -8435,6 +8441,55 @@ function stPlayerInvolvedTurn() {
...
+                break;            
+
+            // id 199, Artifacts age 9: Philips Compact Cassette
+            case "199C1":
+                for($color = 0; $color < 5; $color++) {
+                    self::splay($player_id, $player_id, $color, 0); // Unsplay
+                }
+                break;
+
+            case "199N1":
```
Comment (micahstairs) [inline innovation.game.php:8457]: Okay, so I think "Splay up two colors on your board" can be implemented a lot more simply! First try implementing "Splay up any color on your board", and then repeat the effect a second time (by subtracting one from the step, like Oars).

"Splay up any color on your board" is very similar to the Flight card:

// id 77, age 8: Flight
        case "77N1A":
            // "You may splay any one color of your cards up"
            $options = array(
                'player_id' => $player_id,
                'n' => 1,
                'can_pass' => true,
                
                'splay_direction' => 3 /* up */
            );
            break;

Except in our case, we can't pass!

**Review by micahstairs — CHANGES_REQUESTED** [review]
Thanks for catching that bug! Left some comments.

File: innovation.game.php [hunk 4]
```diff
@@ -8435,6 +8441,55 @@ function stPlayerInvolvedTurn() {
...
+
+            // id 196, Artifacts age 9: Luna 3
+            case "196N1":
+                $step_max = 1;
+                break;            
+
+            // id 199, Artifacts age 9: Philips Compact Cassette
+            case "199C1":
+                for($color = 0; $color < 5; $color++) {
+                    self::splay($player_id, $player_id, $color, 0); // Unsplay
```
[the remark this answers is not in the document]
Comment (ultimatefiend) [reply innovation.game.php:8453]: done

File: innovation.game.php [hunk 5]
```diff
@@ -12664,6 +12787,62 @@ function stInterInteractionStep() {
...
+                    for($card_idx = 0; $card_idx < 5; $card_idx++) { // Put the revealed cards in hand
+                        self::transferCardFromTo($card_array[$card_idx], $player_id, 'hand');
+                    }
+                   
+                    break;                    
+                    
+                case "204N1B":
+                    // "If you have exactly 25 points, you win."
+                    if (self::getPlayerScore($player_id) == 25) {
+                        self::notifyPlayer($player_id, 'log', clienttranslate('${You} have 25 points exactly.'), array('You' => 'You'));
```
[the remark this answers is not in the document]
Comment (ultimatefiend) [reply innovation.game.php:12839]: done

File: innovation.game.php [hunk 6]
```diff
@@ -12664,6 +12787,62 @@ function stInterInteractionStep() {
...
+                        self::notifyAllPlayersBut($player_id, 'log', clienttranslate('${player_name} has 25 points exactly.'), array('player_name' => self::getColoredText(self::getPlayerNameFromId($player_id), $player_id)));
```
[the remark this answers is not in the document]
Comment (ultimatefiend) [reply innovation.game.php:12840]: done

File: innovation.game.php (hunk 2 above)
Comment (ultimatefiend) [reply innovation.game.php:12842]: done

File: innovation.game.php (hunk 1 above)
Comment (ultimatefiend) [reply innovation.game.php:12826]: done

File: innovation.game.php (hunk 3 above)
Comment (ultimatefiend) [reply innovation.game.php:8457]: done

File: innovation.game.php [hunk 7]
```diff
@@ -12786,6 +12904,52 @@ function stInterInteractionStep() {
...
+                        $card = self::executeDraw($player_id, 9, 'revealed');
+                        if ($color_detect[$card['color']] == 1) { 
+                            $repeat_colors = true; // if the flag is 1, the color already has been revealed
+                        }
+                        $color_detect[$card['color']] = 1; // flag the color
+                        $card_array[] = $card;
+                    }
+                    
+                    // If you revealed all five colors, you win.
+                    if ($repeat_colors == false) {
```
Comment (micahstairs) [inline innovation.game.php:12929]: I was able to simplify this a lot by using array_count_values to help count the unique values.

**Review by micahstairs — APPROVED** [review]
Thanks, nice work! I just simplified a few card implementations.

**Pull request merged**